File: app/src/main/java/net/ivpn/client/ui/login/LoginViewModel.java
package net.ivpn.client.ui.login;

import android.content.Context;
import androidx.databinding.ObservableBoolean;
import androidx.databinding.ObservableField;

import com.wireguard.android.crypto.Keypair;

import net.ivpn.client.BuildConfig;
import net.ivpn.client.R;
import net.ivpn.client.common.Mapper;
import net.ivpn.client.common.prefs.ServersRepository;
import net.ivpn.client.common.prefs.Settings;
import net.ivpn.client.common.prefs.UserPreference;
import net.ivpn.client.common.utils.ConnectivityUtil;
import net.ivpn.client.rest.HttpClientFactory;
import net.ivpn.client.rest.RequestListener;
import net.ivpn.client.rest.Responses;
import net.ivpn.client.rest.data.model.WireGuard;
import net.ivpn.client.rest.data.session.SessionNewRequestBody;
import net.ivpn.client.rest.data.session.SessionNewResponse;
import net.ivpn.client.rest.data.wireguard.ErrorResponse;
import net.ivpn.client.rest.requests.common.Request;
import net.ivpn.client.ui.dialog.Dialogs;
import net.ivpn.client.vpn.Protocol;
import net.ivpn.client.vpn.ProtocolController;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.InterruptedIOException;

import javax.inject.Inject;

public class LoginViewModel {

    private static final Logger LOGGER = LoggerFactory.getLogger(LoginViewModel.class);
    private static final String ivpnPrefix = "ivpn";

    public final ObservableField<String> username = new ObservableField<>();
    public final ObservableField<String> usernameError = new ObservableField<>();
    public final ObservableBoolean dataLoading = new ObservableBoolean();

    private Context context;
    private LoginNavigator navigator;
    private Request<SessionNewResponse> request;

    private UserPreference userPreference;
    private ProtocolController protocolController;

    private Settings settings;

    @Inject
    LoginViewModel(Context context, Settings settings, UserPreference userPreference,
                   HttpClientFactory clientFactory, ServersRepository serversRepository,
                   ProtocolController protocolController) {
        this.context = context;
        this.settings = settings;
        this.userPreference = userPreference;
        this.protocolController = protocolController;
        request = new Request<>(settings, clientFactory, serversRepository, Request.Duration.SHORT);

        username.set(userPreference.getUserLogin());
    }

    public void login(boolean force) {
        LOGGER.info("Trying to login");
        if (username.get() == null || !username.get().startsWith(ivpnPrefix)) {
            usernameError.set("Account ID should start with \"ivpn\".");
            return;
        }
        dataLoading.set(true);
        resetErrors();
        login(username.get().trim(), getWgKeyPair(), force);
    }

    public void login(final String username, final Keypair keys, boolean force) {
        String publicKey = keys != null ? keys.getPublicKey() : null;

        SessionNewRequestBody body = new SessionNewRequestBody(username, publicKey, force);
        request.start(api -> api.newSession(body),
                new RequestListener<SessionNewResponse>() {
                    @Override
                    public void onSuccess(SessionNewResponse response) {
                        LOGGER.info("Login process: SUCCESS. Response = " + response);
                        dataLoading.set(false);
                        LoginViewModel.this.onSuccess(username, keys, response);
                    }

                    @Override
                    public void onError(Throwable throwable) {
                        LOGGER.error("Login process: ERROR", throwable);
                        if (!ConnectivityUtil.isOnline(context)) {
                            navigator.openErrorDialogue(Dialogs.CONNECTION_ERROR);
                            dataLoading.set(false);
                            return;
                        }

                        if (throwable instanceof InterruptedIOException) {
                            dataLoading.set(false);
                            navigator.openErrorDialogue(Dialogs.TOO_MANY_ATTEMPTS_ERROR);
                            return;
                        }
                    }

                    @Override
                    public void onError(String error) {
                        LOGGER.error("Login process: ERROR: " + error);
                        ErrorResponse errorResponse = Mapper.errorResponseFrom(error);
                        dataLoading.set(false);
                        handleErrorResponse(errorResponse, username);
                    }
                });
    }

    public void cancel() {
        LOGGER.info("cancel");
        dataLoading.set(false);
        request.cancel();
    }

    public void setNavigator(LoginNavigator navigator) {
        this.navigator = navigator;
    }

    private Keypair getWgKeyPair() {
        Protocol currentProtocol = protocolController.getCurrentProtocol();
        if (currentProtocol.equals(Protocol.WIREGUARD)) {
            return new Keypair();
        }
        return null;
    }

    private void onSuccess(String username, Keypair keys, SessionNewResponse response) {
        if (response.getStatus() == null) {
            navigator.openErrorDialogue(Dialogs.SERVER_ERROR);
            return;
        }

        LOGGER.info("Status = " + response.getStatus());
        if (response.getStatus() == Responses.SUCCESS) {
            putUserData(username, response);
            handleWireGuardResponse(keys, response.getWireGuard());
            if (userPreference.getIsActive()) {
                navigator.onLogin();
            } else {
                if (BuildConfig.BUILD_VARIANT.equals("site")) {
                    navigator.openSite();
                } else {
                    navigator.openSubscriptionScreen();
                }
            }
        } else {
            navigator.openErrorDialogue(Dialogs.SERVER_ERROR);
        }
    }

    private void handleErrorResponse(ErrorResponse errorResponse, String username) {
        if (errorResponse == null || errorResponse.getStatus() == null) {
            navigator.openErrorDialogue(Dialogs.SERVER_ERROR);
            return;
        }

        switch (errorResponse.getStatus()) {
            case Responses.INVALID_CREDENTIALS: {
                navigator.openErrorDialogue(Dialogs.AUTHENTICATION_ERROR);
                break;
            }
            case Responses.NOT_ACTIVE: {
                userPreference.putUserLogin(username);
                navigator.openActivateDialogue();
                break;
            }
            case Responses.SESSION_TOO_MANY: {
                navigator.openSessionLimitReachedDialogue();
                break;
            }

            case Responses.WIREGUARD_KEY_INVALID:
            case Responses.WIREGUARD_PUBLIC_KEY_EXIST:
            case Responses.BAD_REQUEST:
            case Responses.SESSION_SERVICE_ERROR:
            default: {
                navigator.openCustomErrorDialogue(context.getString(R.string.dialogs_error) + errorResponse.getStatus(),
                        errorResponse.getMessage() != null ? errorResponse.getMessage() : "");
                break;
            }
        }
    }

    private void handleWireGuardResponse(Keypair keys, WireGuard wireGuard) {
        if (wireGuard == null || wireGuard.getStatus() == null) {
            resetWireGuard();
            return;
        }

        if (wireGuard.getStatus() == Responses.SUCCESS) {
            putWireGuardData(keys, wireGuard);
        } else {
            LOGGER.error("Error received: " + wireGuard.getStatus() + " "
                    + (wireGuard.getMessage() != null ? wireGuard.getMessage() : ""));
            resetWireGuard();
        }
    }

    private void putUserData(String username, SessionNewResponse response) {
        LOGGER.info("Save account data");

        userPreference.putSessionToken(response.getToken());
        userPreference.putSessionUsername(response.getVpnUsername());
        userPreference.putSessionPassword(response.getVpnPassword());

        userPreference.putAvailableUntil(response.getServiceStatus().getActiveUntil());
        userPreference.putIsUserOnTrial(Boolean.valueOf(response.getServiceStatus().getIsOnFreeTrial()));
        userPreference.putCurrentPlan(response.getServiceStatus().getCurrentPlan());
        userPreference.putPaymentMethod(response.getServiceStatus().getPaymentMethod());
        userPreference.putIsActive(response.getServiceStatus().getIsActive());

        if (response.getServiceStatus().getCapabilities() != null) {
            userPreference.putIsUserOnPrivateEmailBeta(response.getServiceStatus().getCapabilities().contains(Responses.PRIVATE_EMAILS));
            boolean multiHopCapabilities = response.getServiceStatus().getCapabilities().contains(Responses.MULTI_HOP);
            userPreference.putCapabilityMultiHop(response.getServiceStatus().getCapabilities().contains(Responses.MULTI_HOP));
            if (!multiHopCapabilities) {
                settings.enableMultiHop(false);
            }
        }

        userPreference.putUserLogin(username);
    }

    private void resetErrors() {
        usernameError.set(null);
    }

    private void putWireGuardData(Keypair keys, WireGuard wireGuard) {
        settings.saveWireGuardKeypair(keys);
        settings.setWireGuardIpAddress(wireGuard.getIpAddress());
    }

    private void resetWireGuard() {
        protocolController.setCurrentProtocol(Protocol.OPENVPN);
    }
}

File: app/src/main/java/net/ivpn/client/v2/login/LoginViewModel.kt
package net.ivpn.client.v2.login

/*
 IVPN Android app
 https://github.com/ivpn/android-app

 Created by Oleksandr Mykhailenko.
 Copyright (c) 2020 Privatus Limited.

 This file is part of the IVPN Android app.

 The IVPN Android app is free software: you can redistribute it and/or
 modify it under the terms of the GNU General Public License as published by the Free
 Software Foundation, either version 3 of the License, or (at your option) any later version.

 The IVPN Android app is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License
 along with the IVPN Android app. If not, see <https://www.gnu.org/licenses/>.
*/

import android.content.Context
import android.view.View
import androidx.core.text.isDigitsOnly
import androidx.databinding.ObservableBoolean
import androidx.databinding.ObservableField
import androidx.lifecycle.ViewModel
import net.ivpn.client.R
import net.ivpn.client.common.dagger.ApplicationScope
import net.ivpn.client.common.prefs.EncryptedUserPreference
import net.ivpn.client.common.session.SessionController
import net.ivpn.client.common.session.SessionListenerImpl
import net.ivpn.client.common.utils.ConnectivityUtil
import net.ivpn.client.rest.Responses
import net.ivpn.client.rest.data.session.SessionNewResponse
import net.ivpn.client.rest.data.wireguard.ErrorResponse
import net.ivpn.client.v2.dialog.Dialogs
import org.slf4j.LoggerFactory
import java.io.InterruptedIOException
import javax.inject.Inject

@ApplicationScope
class LoginViewModel @Inject constructor(
        private val context: Context,
        private val userPreference: EncryptedUserPreference,
        private val sessionController: SessionController
) : ViewModel() {

    companion object {
        private const val ivpnPrefix = "ivpn"
        private const val ivpnNewPrefix = "i-"
        private val LOGGER = LoggerFactory.getLogger(LoginViewModel::class.java)
    }

    val username = ObservableField<String>()
    val error = ObservableField<String>()

    val captchaInputState = ObservableField<InputState>()
    val tfaInputState = ObservableField<InputState>()
    val loginInputState = ObservableField<InputState>()

    val tfaToken = ObservableField<String>()
    val captcha = ObservableField<String>()
    var captchaId: String? = null
    var captchaImage = ObservableField<String>()

    val dataLoading = ObservableBoolean()

    var paymentMethod: String? = null

    var navigator: LoginNavigator? = null

    var pendingCaptcha: String? = null
    var pending2FAToken: String? = null

    var tfaFocusListener = View.OnFocusChangeListener { _, hasFocus ->
        if (hasFocus) {
            tfaInputState.get()?.let {
                if (it == InputState.NORMAL) {
                    tfaInputState.set(InputState.FOCUSED)
                }
            } ?: kotlin.run {
                tfaInputState.set(InputState.FOCUSED)
            }
        } else {
            tfaInputState.set(InputState.NORMAL)
        }
    }

    var loginFocusListener = View.OnFocusChangeListener { _, hasFocus ->
        if (hasFocus) {
            loginInputState.get()?.let {
                if (it == InputState.NORMAL) {
                    loginInputState.set(InputState.FOCUSED)
                }
            } ?: kotlin.run {
                loginInputState.set(InputState.FOCUSED)
            }
        } else {
            loginInputState.set(InputState.NORMAL)
        }
    }

    val captchaFocusListener = View.OnFocusChangeListener { _, hasFocus ->
        if (hasFocus) {
            captchaInputState.get()?.let {
                if (it == InputState.NORMAL) {
                    captchaInputState.set(InputState.FOCUSED)
                }
            } ?: kotlin.run {
                captchaInputState.set(InputState.FOCUSED)
            }
        } else {
            captchaInputState.set(InputState.NORMAL)
        }
    }

    init {
        dataLoading.set(false)
        sessionController.subscribe(object : SessionListenerImpl() {
            override fun onCreateSuccess(response: SessionNewResponse) {
                LOGGER.info("Login process: SUCCESS. Response = $response")
                dataLoading.set(false)
                this@LoginViewModel.onSuccess(response)
            }

            override fun onCreateError(throwable: Throwable?, errorResponse: ErrorResponse?) {
                LOGGER.error("Login process: ERROR", throwable)
                if (!ConnectivityUtil.isOnline(context)) {
                    navigator?.openErrorDialogue(Dialogs.CONNECTION_ERROR)
                    dataLoading.set(false)
                    return
                }
                if (throwable is InterruptedIOException) {
                    dataLoading.set(false)
                    navigator?.openErrorDialogue(Dialogs.TOO_MANY_ATTEMPTS_ERROR)
                    return
                }

                LOGGER.error("Login process: ERROR: $errorResponse")
                dataLoading.set(false)
                handleErrorResponse(errorResponse)
            }
        })

        username.set(userPreference.getUserLogin())
    }

    fun onResume() {
        paymentMethod = getPaymentMethodValue()
    }

    fun login(force: Boolean) {
        LOGGER.info("Trying to login")
        username.get()?.let {
            if (!(it.startsWith(ivpnPrefix) || it.startsWith(ivpnNewPrefix))) {
                loginInputState.set(InputState.ERROR)
                error.set("Your account ID has to be in 'i-XXXX-XXXX-XXXX' or 'ivpnXXXXXXXX'" +
                        " format. You can find it on other devices where you are logged in and in the" +
                        " client area of the IVPN website.")
                return
            }
        } ?: kotlin.run {
            loginInputState.set(InputState.ERROR)
            return
        }

        dataLoading.set(true)
        resetErrors()
        username.get()?.let { usernameObj ->
            pendingCaptcha?.let { captchaObj ->
                captchaId?.let { captchaIdObj ->
                    sessionController.createSessionWithCaptcha(force, usernameObj.trim(), captchaIdObj, captchaObj)
                    pendingCaptcha = null
                    return
                }
            }
            pending2FAToken?.let { tfaTokenObj ->
                sessionController.createSessionWith2FAToken(force, usernameObj.trim(), tfaTokenObj)
                pending2FAToken = null
                return
            }
            login(usernameObj.trim(), force)
        }
    }

    fun submit2FAToken() {
        LOGGER.info("submit2FAToken")
        tfaToken.get()?.let {
            if (it.length != 6 || !it.isDigitsOnly()) {
                error.set("Please enter 6-digit verification code")
                tfaInputState.set(InputState.ERROR)
                return
            }
            username.get()?.let { usernameObj ->
                dataLoading.set(true)
                sessionController.createSessionWith2FAToken(false, usernameObj, it)
            }
            resetErrors()
        } ?: kotlin.run {
            error.set("Please enter 6-digit verification code")
            tfaInputState.set(InputState.ERROR)
        }
    }

    fun submitCaptcha() {
        LOGGER.info("submitCaptcha navigator = $navigator")
        captcha.get()?.let {
            if (it.isEmpty()) {
                error.set("Please enter 6-digit verification code")
                captchaInputState.set(InputState.ERROR)
                return
            }
            username.get()?.let { usernameObj ->
                captchaId?.let { captchaIdObj ->
                    dataLoading.set(true)
                    sessionController.createSessionWithCaptcha(false, usernameObj, captchaIdObj, it)
                }
            }
            resetErrors()
        } ?: kotlin.run {
            error.set("Please enter 6-digit verification code")
            captchaInputState.set(InputState.ERROR)
        }
    }

    fun login(username: String, force: Boolean) {
        sessionController.createSession(force, username)
    }

    fun isAccountNewStyle(): Boolean {
        return paymentMethod?.let {
            it == "prepaid"
        } ?: false
    }

    fun getAccountType(): String? {
        return userPreference.getCurrentPlan()
    }

    fun cancel() {
        LOGGER.info("cancel")
        dataLoading.set(false)
        sessionController.cancel()
    }

    private fun onSuccess(response: SessionNewResponse) {
        response.status?.let {
            if (it != Responses.SUCCESS) {
                navigator?.openErrorDialogue(Dialogs.SERVER_ERROR)
                return
            }
        } ?: return

        LOGGER.info("Status = ${response.status}")
        username.get()?.let { accountId ->
            if (accountId.isNotEmpty()) {
                userPreference.putUserLogin(accountId)
            }
        }
        if (userPreference.getIsActive()) {
            navigator?.onLogin()
        } else {
            navigator?.onLoginWithInactiveAccount()
        }
    }

    private fun handleErrorResponse(errorResponse: ErrorResponse?) {
        errorResponse?.let {
            if (it.status == null) {
                navigator?.openErrorDialogue(Dialogs.SERVER_ERROR)
                return
            }
        } ?: kotlin.run {
            navigator?.openErrorDialogue(Dialogs.CREATE_SESSION_ERROR)
            return
        }

        when (errorResponse.status) {
            Responses.ACCOUNT_NOT_ACTIVE -> {
                navigator?.onLoginWithBlankAccount()
            }
            Responses.ENTER_TOTP_TOKEN -> {
                navigator?.openTFAScreen()
            }
            Responses.INVALID_TOTP_TOKEN -> {
                tfaToken.set(null)
                error.set("Specified two-factor authentication token is not valid")
                tfaInputState.set(InputState.ERROR)
            }
            Responses.INVALID_CREDENTIALS -> {
                navigator?.onInvalidAccount()
                navigator?.openErrorDialogue(Dialogs.AUTHENTICATION_ERROR)
            }
            Responses.SESSION_TOO_MANY -> {
                pendingCaptcha = captcha.get()
                pending2FAToken = tfaToken.get()
                navigator?.openSessionLimitReachedDialogue()
            }
            Responses.INVALID_CAPTCHA -> {
                error.set("Invalid captcha, please try again")
                captchaInputState.set(InputState.ERROR)
                captcha.set(null)
                captchaId = errorResponse.captchaId
                captchaImage.set(errorResponse.captchaImage)
                println("captchaId = $captchaId")
                println("captchaImage = $captchaImage")
                navigator?.openCaptcha()
            }
            Responses.CAPTCHA_REQUIRED -> {
                captchaId = errorResponse.captchaId
                captchaImage.set(errorResponse.captchaImage)
                println("captchaId = $captchaId")
                println("captchaImage = $captchaImage")
                navigator?.openCaptcha()
            }
            Responses.WIREGUARD_KEY_INVALID, Responses.WIREGUARD_PUBLIC_KEY_EXIST, Responses.BAD_REQUEST, Responses.SESSION_SERVICE_ERROR -> {
                navigator?.openCustomErrorDialogue("${context.getString(R.string.dialogs_error)} ${errorResponse.status}",
                        if (errorResponse.message != null) errorResponse.message else "")
            }
            else -> {
                navigator?.openCustomErrorDialogue("${context.getString(R.string.dialogs_error)} ${errorResponse.status}",
                        if (errorResponse.message != null) errorResponse.message else "")
            }
        }
    }

    fun reset() {
        error.set(null)
        captcha.set(null)
        tfaToken.set(null)
        captchaInputState.set(InputState.NORMAL)
        tfaInputState.set(InputState.NORMAL)
        loginInputState.set(InputState.NORMAL)
    }

    private fun resetErrors() {
        error.set(null)
        captchaInputState.set(InputState.NORMAL)
        tfaInputState.set(InputState.NORMAL)
        loginInputState.set(InputState.NORMAL)
    }

    private fun getPaymentMethodValue(): String {
        return userPreference.getPaymentMethod()
    }

    enum class InputState {
        NORMAL,
        FOCUSED,
        ERROR
    }

}